### ISSUE TITLE
IScheduledExecutorService always schedules a distributed task on the same Node where it first ran

### ISSUE THREAD

**Issue by RameshP96** [hazelcast/hazelcast#16346]
We have an application(embeds hz) and acts as a cluster-node. We use **IScheduledExecutorService** to schedule/run tasks across the cluster. 
I see the task-execution behavior of IScheduledExecutorService is very specific, is there any way to change that behavior? 

To explain better:

### Problem 1: 
IScheduledExecutorService.scheduleAtFixedRate() API  always runs a task on same Node
--> Suppose i start/run 3 instances of our application (i.e. basically 3 hz nodes cluster)
--> After all the application-nodes are started, Supposing 6 tasks are created and scheduled for every 10 minutes (i.e. using scheduleAtFixedRate API)
       Supposing **best scenario** --> Hz distributes those 6 tasks evenly across 3 nodes
                      Node1 --> task1, task3
                      Node2 --> task2, task6
                      Node3 --> task6, task4
         Once the first time a task is run on a Node, that task would always be run on the same Node until that Node is available or another Node is added/removed. 
         If there is no change in the cluster in terms of Nodes, the task continue on the same Node
   
 _is there any way  IScheduledExecutorService can decide/run a task on different nodes for each run based on OS-resource availability on Nodes at just before running?_

_I want for instance, task1 to run one-time on Node1, secondtime on some other node and third time on another node and so forth, same is the case with other tasks as well._

### Problem 2:
Some times i see the tasks are not distributed evenly across Nodes, so few Nodes will continue to be free where as the Nodes to which the tasks are assigned will continue to be busy

    Sometimes i see the tasks are not evenly distributed across the nodes
                      Node1 --> task1, task3, task4,
                      Node2 --> task2, task6, task6
                      Node3 --> <no task>
        Or
                      Node1 --> task1, task3, task4, task6
                      Node2 --> task2, task6
                      Node3 --> <no task>   

    _With this the problem is the tasks that were assigned initially to a Node (may be based on the then resource availability) will continue run on the same Node._
   _So as per above the Node3 will never get to run any task and will be free until/unless there is a change in the cluster nodes (addition/removal) and Node1/Node2 will be busy_
    _is there a way to govern IScheduledExecutorService to distribute tasks evenly?_

### Problem 3: 
     Are there plans to support **scheduleWithFixedDelay** with IScheduledExecutorService similar to ScheduledExecutorService?
     I don't want to run costly tasks in fixed-interval to hit external-servers, instead i would like to give breathing time for external-servers so that they wouldn't get overloading. For instance, if a task is scheduled every 10 seconds and in _worst-case if the task is going to take 8/9 seconds it is going to overload external-server to run that task kind of continuous,_

These are some of the pains i am experiencing with IScheduledExecutorService behavior, in any case 

**Hazelcast is really a great library, hats off to the team behind it and am loving it**

Thanks in advance
Ramesh

**Comment by tkountis** [comment]
Hi @RameshP96 thanks for your kind words for Hazelcast :100: 

Regarding your request, I think there might be some slight misconception on your understanding, of how the Executor distributes tasks. What really happens, when you use that API, is that we randomly create a unique name for your task, and then based on that unique name we pick the **matching** partition to assign it. This decision has nothing to do with resource availability, but its rather based on hashing that name on slot on that partition table. As you correctly noted, this slot can be re-assigned to another node in the cluster when the cluster changes.

So to make this more clear, the real owner of the task is a partition, as with most of Hazelcast's data structures. Nodes are assigned partitions, and that changes only when the cluster layout changes anyhow. Until that happens, a task will always execute on the same node, owning that partition.

Now, with that in mind, the answers to your problems.

**Problem 1**
Based on what we said, there is no partition relocation done, so the task will indeed execute on that node until the cluster layout changes. This is not something we have in the roadmap to change. However, I believe you could have a workaround, by having a governor task that can cancel and re-schedule a given task on different node, by using any of the 'NamedTask' or 'PartitionAware' marker interfaces on your tasks.

**Problem 2**
The task distribution is happening based on a random name generator on the tasks. If you find that distribution to not be even, you can always, once again, rely on 'NamedTask' or 'PartitionAware' marker interfaces on your tasks to decide on which partition they run.

**Problem 3**
That is indeed not supported at this stage, I can imagine of a workaround on your end, by keeping track of execution time, and deciding whether you should run next time or not, but its rather complex to implement, so I wont even go through the trouble of suggesting it :-)
There is no immediate plat to support it, but I would recommend that once we resolve the other problem in this issue, we close it, and open a separate request to add support for that functionality. Once it gets some more community love, it will certainly get prioritized.

Once again thank you for your kind words, please let me know if any of the above helped, or if you need any further guidance.

**Comment by RameshP96** [comment]
Hi Thomas,
Thank you very much for explaining me about how the tasks are distributed by Hazelcast internally (i.e. not based on system-resources rather based on the generated hash of task-name). I also have gone through the NamedTask, PartitionAware interfaces and got some idea.

Regarding **Problem3**, I am fine for now but I will plan to open a separate request as suggested. 
I have few more followup questions related to **Problem1** and **Problem2**

_For these 2 usecases, I am not in favor of using NamedTask, as I don't want to change each time the task-name "somehow to map-to-a-partition" in order to move/run task each time on different Node_
**So in relation to "PartitionAware" interface ->** When does the PartitionAware.getPartitionKey() is called for a task upon calling IScheduledExecutorService.scheduleAtFixedRate() API? 
i) is it only first time + every partition-change time only? 
ii) or each time, just before running a scheduled-task?

You have also suggested a governor task to cancel and reschedule the task -> is this because of the "i)" above? if yes, is it costly to cancel/reschedule a task by something like governor?

If it is "ii)" above, then using appropriate code in PartitionAware.getPartitionKey() implementation to return each time a different nodes partitionId would do without needing a governor task, is that so?

I also referred HazelcastInstanceObj.getPartitionService() to see how to effectively get one partitionkey of each owner-member but I didn't find other than going through **HazelcastInstanceObj.getPartitionService().getPartitions()** API and then iterating over until finding one partitionkey for each owner, **is this the only way**?

**Additional questions**: 
1) is there a way to **get all HazelcastInstance objects in a/total cluster** (analogous to  Hazelcast.getAllHazelcastInstances() API which gives per JVM)? 
Basically the requirement is that, i **don't want to schedule few special tasks on specific Hz-Nodes** because **those Nodes don't understand the "special" tasks (if run, tasks would fail on them)** so i want to use something like ((PartitionAware)SpecialTask).getPartitionKey() API to return partition-key of one of the AllowedNodes each time (i.e. checking UserContext of each HazelcastInstanceObj to identify which ones are AllowedNodes?).
2) How do I get HazelcastInstanceObject from a Member object? especially when inside a Listener execution? 

Thanks in advance
Ramesh

**Comment by tkountis** [comment]
Hi @RameshP96 

My answers inline

>Regarding Problem3, I am fine for now but I will plan to open a separate request as suggested.
I have few more followup questions related to Problem1 and Problem2
>
>For these 2 usecases, I am not in favor of using NamedTask, as I don't want to change each time the task-name "somehow to map-to-a-partition" in order to move/run task each time on different Node
So in relation to "PartitionAware" interface -> When does the PartitionAware.getPartitionKey() is called for a task upon calling IScheduledExecutorService.scheduleAtFixedRate() API?
i) is it only first time + every partition-change time only?
ii) or each time, just before running a scheduled-task?

This happens upon scheduling on the *caller* side, the very first time. Once the partition id is assigned it sticks for the remaining of the life of that task.

>You have also suggested a governor task to cancel and reschedule the task -> is this because of the "i)" above? if yes, is it costly to cancel/reschedule a task by something like governor?

Correct, for the above reason, the work around would be to have another light-weight governor task, which can cancel and re-schedule the tasks according to your needs. I think it would even be more reliable, if the tasks where not repetitive but rather single-runs, and have the governor manage their repetition. This way, you can easier guarantee, when the task finished running before re-scheduling on the other node. This recommendation is mostly due to the weak semantic of the `Future.cancel()` which can not guarantee that current execution is cancelled.

>I also referred HazelcastInstanceObj.getPartitionService() to see how to effectively get one partitionkey of each owner-member but I didn't find other than going through HazelcastInstanceObj.getPartitionService().getPartitions() API and then iterating over until finding one partitionkey for each owner, is this the only way?

We don't have any public API to create keys for specific members/partitions, you could have a look at [this code](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java#L620) here, you could mimic what we do.

>is there a way to get all HazelcastInstance objects in a/total cluster (analogous to Hazelcast.getAllHazelcastInstances() API which gives per JVM)?
Basically the requirement is that, i don't want to schedule few special tasks on specific Hz-Nodes because those Nodes don't understand the "special" tasks (if run, tasks would fail on them) so i want to use something like ((PartitionAware)SpecialTask).getPartitionKey() API to return partition-key of one of the AllowedNodes each time (i.e. checking UserContext of each HazelcastInstanceObj to identify which ones are AllowedNodes?).

> How do I get HazelcastInstanceObject from a Member object? especially when inside a Listener execution?

If I understand the questions correctly, then the answer to both of them, is to tag your runnables/callables as `HazelcastInstanceAware` which will inject the `HazelcastInstace` when they run, so you can access instance APIs.

Let me know if I misunderstood something, or missed anything.

**Comment by tkountis** [comment]
Hi @RameshP96, is there any update on your progress here? 
I will close the issue for now, since its been inactive for a couple of weeks. If you need further assistance please free to reach out to https://gitter.im/hazelcast/hazelcast or if you feel that you are still facing issues with the API, you can re-open the issue anytime.
